In [177]:
import numpy as np
import pandas as pd
import matplotlib as plt
import matplotlib.pyplot as plt
import seaborn as sns

In [178]:
# null hypothesis- one categorical
# from scipy import stats

# F, p = stats.f_oneway(red_wine[red_wine['quality_label'] == 'low']['alcohol'],
# red_wine[red_wine['quality_label'] == 'medium']['alcohol'],
# red_wine[red_wine['quality_label'] == 'high']['alcohol'])
# print('ANOVA test for mean alcohol levels across wine samples with different quality ratings')
# print('F Statistic:', F, '\tp-value:', p)

## Data Preparatrion

In [179]:
white_wine_heatmap = pd.read_csv('winequality-white.csv', sep=';')

In [180]:
white_wine_heatmap

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0              0.27         0.36            20.7      0.045   
1               6.3              0.30         0.34             1.6      0.049   
2               8.1              0.28         0.40             6.9      0.050   
3               7.2              0.23         0.32             8.5      0.058   
4               7.2              0.23         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
4893            6.2              0.21         0.29             1.6      0.039   
4894            6.6              0.32         0.36             8.0      0.047   
4895            6.5              0.24         0.19             1.2      0.041   
4896            5.5              0.29         0.30             1.1      0.022   
4897            6.0              0.21         0.38             0.8      0.020   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
4893                 24.0                  92.0  0.99114  3.27       0.50   
4894                 57.0                 168.0  0.99490  3.15       0.46   
4895                 30.0                 111.0  0.99254  2.99       0.46   
4896                 20.0                 110.0  0.98869  3.34       0.38   
4897                 22.0                  98.0  0.98941  3.26       0.32   

      alcohol  quality  
0         8.8        6  
1         9.5        6  
2        10.1        6  
3         9.9        6  
4         9.9        6  
...       ...      ...  
4893     11.2        6  
4894      9.6        5  
4895      9.4        6  
4896     12.8        7  
4897     11.8        6  

[4898 rows x 12 columns]

In [181]:
white_wine_heatmap.drop_duplicates(inplace=True)
white_wine_heatmap.reset_index(inplace=True)
white_wine_heatmap.drop(columns='index', inplace=True)

In [182]:
white_wine_heatmap.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3961 entries, 0 to 3960
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         3961 non-null   float64
 1   volatile acidity      3961 non-null   float64
 2   citric acid           3961 non-null   float64
 3   residual sugar        3961 non-null   float64
 4   chlorides             3961 non-null   float64
 5   free sulfur dioxide   3961 non-null   float64
 6   total sulfur dioxide  3961 non-null   float64
 7   density               3961 non-null   float64
 8   pH                    3961 non-null   float64
 9   sulphates             3961 non-null   float64
 10  alcohol               3961 non-null   float64
 11  quality               3961 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 371.5 KB


### add quality label

In [183]:
# we are creating a new column called "quality_label", we define a range and associate that range with a label
white_wine_heatmap['quality label'] = white_wine_heatmap['quality'].apply(lambda value: 'low'
if value <= 5 else 'medium'
if value <= 7 else 'high')

# here we are transforming these labels into categrical data type (specific to pandas) instead of simple string
white_wine_heatmap['quality label'] = pd.Categorical(white_wine_heatmap['quality label'],
categories=['low', 'medium', 'high'])

# hypothesis check

In [184]:
from scipy import stats

In [185]:
def anova_tests2(df):  
    
    features = ['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar', 'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol']
    
    for feature in features:
        F, p = stats.f_oneway(
            df[df['quality label'] == 'low'][feature],
            df[df['quality label'] == 'medium'][feature],
            df[df['quality label'] == 'high'][feature]
        )
        if p > 0.05:
         print(f"ANOVA test for mean {feature} across wine samples with different quality ratings")
         print('F Statistic:', F, '\tp-value:', p)
        else:
         print("there is correlation")

In [186]:
anova_tests2(white_wine_heatmap)


there is correlation
there is correlation
ANOVA test for mean citric acid across wine samples with different quality ratings
F Statistic: 0.24868681586895755 	p-value: 0.7798363477964527
there is correlation
there is correlation
ANOVA test for mean free sulfur dioxide across wine samples with different quality ratings
F Statistic: 0.11569157788750334 	p-value: 0.8907529199895233
there is correlation
there is correlation
there is correlation
there is correlation
there is correlation


In [187]:
# changing wine type to 0 and 1
wine1=white_wine_heatmap.copy()

In [188]:
wine1.drop(columns='citric acid', inplace=True)

In [189]:
wine1.drop(columns='free sulfur dioxide', inplace=True)

## encoding and dropping

In [190]:
from sklearn.preprocessing import OrdinalEncoder
# Create an instance of OrdinalEncoder

In [191]:
ordinal_encoder = OrdinalEncoder(categories=[['low', 'medium', 'high']])
wine1['NumericalLabels'] = ordinal_encoder.fit_transform(wine1[['quality label']])

In [192]:
wine1.drop(columns='quality', inplace=True)

In [193]:
wine1.drop(columns='quality label', inplace=True)

# Models

divide into train and test

In [194]:
#importm library for dicidint the dataset into two 
from sklearn.model_selection import train_test_split

In [195]:
wine1.head(1)

fixed acidity  volatile acidity  residual sugar  chlorides  \
0            7.0              0.27            20.7      0.045   

   total sulfur dioxide  density   pH  sulphates  alcohol  NumericalLabels  
0                 170.0    1.001  3.0       0.45      8.8              1.0

In [196]:
wine2=wine1.copy()

In [197]:
wine2.drop(columns='NumericalLabels', inplace=True)

In [198]:
X=wine2

In [199]:

y = wine1['NumericalLabels']  # --> what you're trying to predict
  # --> the features we will keep to build our model





In [200]:
X.shape

(3961, 9)

In [201]:
y.shape

(3961,)

In [202]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2, random_state=100)

In [203]:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(3168, 9) (793, 9) (3168,) (793,)


## oversampling

In [204]:
from imblearn.over_sampling import SMOTE

In [205]:
from sklearn.datasets import make_classification
from collections import Counter

In [206]:
# Display class distribution before oversampling
print("Class Distribution Before SMOTE (Training Set):", Counter(y_train))

Class Distribution Before SMOTE (Training Set): Counter({1.0: 1987, 0.0: 1074, 2.0: 107})


In [207]:
# Apply SMOTE for oversampling only on the training set
smote = SMOTE(sampling_strategy='auto', random_state=42)
X_train_resampled, y_train_resampled = smote.fit_resample(X_train, y_train)

In [208]:
# Display class distribution after oversampling
print("Class Distribution After SMOTE (Training Set):", Counter(y_train_resampled))

Class Distribution After SMOTE (Training Set): Counter({1.0: 1987, 0.0: 1987, 2.0: 1987})


## Random Forest model

In [209]:
from sklearn.ensemble import RandomForestClassifier


In [210]:
RF_model= RandomForestClassifier(n_estimators= 100, criterion='entropy', random_state=445)

In [211]:
RF_model.fit(X_train_resampled, y_train_resampled)

RandomForestClassifier(criterion='entropy', random_state=445)

In [212]:
preds2= RF_model.predict(X_test)

In [213]:
preds2

array([0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 1., 0., 1., 0., 1., 1.,
       1., 0., 0., 1., 1., 1., 1., 1., 2., 1., 0., 1., 2., 1., 1., 0., 2.,
       0., 1., 1., 1., 1., 0., 0., 2., 1., 0., 0., 1., 0., 1., 1., 1., 0.,
       1., 0., 1., 1., 0., 1., 1., 0., 2., 1., 0., 1., 2., 1., 1., 0., 1.,
       0., 1., 1., 1., 1., 1., 1., 1., 0., 1., 0., 1., 1., 1., 1., 0., 0.,
       0., 1., 0., 1., 1., 0., 0., 1., 1., 0., 0., 1., 1., 1., 1., 1., 0.,
       1., 0., 0., 2., 1., 0., 0., 1., 0., 1., 2., 1., 1., 0., 1., 1., 1.,
       0., 1., 1., 1., 0., 1., 1., 1., 0., 1., 1., 1., 1., 0., 1., 1., 0.,
       1., 0., 1., 1., 1., 0., 0., 0., 1., 1., 1., 0., 0., 2., 1., 1., 1.,
       1., 1., 1., 0., 1., 1., 1., 1., 0., 1., 0., 0., 0., 0., 1., 2., 1.,
       1., 1., 1., 1., 1., 0., 1., 0., 1., 2., 1., 0., 0., 1., 0., 0., 0.,
       2., 2., 0., 1., 0., 1., 1., 1., 1., 1., 0., 1., 0., 0., 0., 2., 1.,
       0., 1., 0., 2., 1., 0., 1., 1., 0., 2., 2., 0., 1., 0., 0., 0., 2.,
       1., 1., 0., 0., 1.

In [214]:
acc=accuracy_score(y_test, preds2)
acc

0.6620428751576293

## results

In [215]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [216]:
accuracy = accuracy_score(y_test, preds2)
print(f"Accuracy: {accuracy}")

print("\nClassification Report:")
print(classification_report(y_test, preds2))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, preds2))

Accuracy: 0.6620428751576293

Classification Report:
              precision    recall  f1-score   support

         0.0       0.60      0.70      0.65       274
         1.0       0.77      0.66      0.71       490
         2.0       0.17      0.31      0.22        29

    accuracy                           0.66       793
   macro avg       0.51      0.56      0.53       793
weighted avg       0.69      0.66      0.67       793


Confusion Matrix:
[[191  79   4]
 [124 325  41]
 [  1  19   9]]


1st try no scaling or oversampling
Accuracy: 0.7301387137452712

Classification Report:
              precision    recall  f1-score   support

         0.0       0.67      0.63      0.65       274
         1.0       0.76      0.83      0.79       490
         2.0       0.50      0.03      0.06        29

    accuracy                           0.73       793
   macro avg       0.64      0.50      0.50       793
weighted avg       0.72      0.73      0.72       793


Confusion Matrix:
[[172 102   0]
 [ 83 406   1]
 [  0  28   1]]

2nd try oversampling:
Accuracy: 0.669609079445145

Classification Report:
              precision    recall  f1-score   support

         0.0       0.61      0.70      0.65       274
         1.0       0.77      0.68      0.72       490
         2.0       0.12      0.17      0.14        29

    accuracy                           0.67       793
   macro avg       0.50      0.52      0.50       793
weighted avg       0.69      0.67      0.68       793


Confusion Matrix:
[[193  79   2]
 [121 333  36]
 [  2  22   5]]

3rd try removed irrelevant columns:
Accuracy: 0.6620428751576293

Classification Report:
              precision    recall  f1-score   support

         0.0       0.60      0.70      0.65       274
         1.0       0.77      0.66      0.71       490
         2.0       0.17      0.31      0.22        29

    accuracy                           0.66       793
   macro avg       0.51      0.56      0.53       793
weighted avg       0.69      0.66      0.67       793


Confusion Matrix:
[[191  79   4]
 [124 325  41]
 [  1  19   9]]


## KNN

In [217]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [218]:
scaler = StandardScaler()
X_train1 = scaler.fit_transform(X_train_resampled)
X_test1 = scaler.transform(X_test)

In [219]:
k = 3  # You can adjust this value based on your problem
knn_model = KNeighborsClassifier(n_neighbors=k)
knn_model.fit(X_train1, y_train_resampled)

KNeighborsClassifier(n_neighbors=3)

## results


In [220]:
y_pred = knn_model.predict(X_test1)

In [221]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

Accuracy: 0.562421185372005

Classification Report:
              precision    recall  f1-score   support

         0.0       0.54      0.66      0.60       274
         1.0       0.74      0.52      0.61       490
         2.0       0.09      0.38      0.15        29

    accuracy                           0.56       793
   macro avg       0.46      0.52      0.45       793
weighted avg       0.65      0.56      0.59       793


Confusion Matrix:
[[180  76  18]
 [147 255  88]
 [  4  14  11]]


1st try
Accuracy: 0.6973518284993695

Classification Report:
              precision    recall  f1-score   support

         0.0       0.62      0.62      0.62       274
         1.0       0.75      0.77      0.76       490
         2.0       0.25      0.14      0.18        29

    accuracy                           0.70       793
   macro avg       0.54      0.51      0.52       793
weighted avg       0.69      0.70      0.69       793


Confusion Matrix:
[[171 101   2]
 [102 378  10]
 [  2  23   4]]

2nd try:
Accuracy: 0.6015132408575031

Classification Report:
              precision    recall  f1-score   support

         0.0       0.59      0.74      0.66       274
         1.0       0.78      0.54      0.64       490
         2.0       0.07      0.24      0.10        29

    accuracy                           0.60       793
   macro avg       0.48      0.51      0.47       793
weighted avg       0.69      0.60      0.63       793


Confusion Matrix:
[[204  59  11]
 [136 266  88]
 [  5  17   7]]

3rd try:

Accuracy: 0.562421185372005

Classification Report:
              precision    recall  f1-score   support

         0.0       0.54      0.66      0.60       274
         1.0       0.74      0.52      0.61       490
         2.0       0.09      0.38      0.15        29

    accuracy                           0.56       793
   macro avg       0.46      0.52      0.45       793
weighted avg       0.65      0.56      0.59       793


Confusion Matrix:
[[180  76  18]
 [147 255  88]
 [  4  14  11]]

## Support Vector Machines (SVM)

In [222]:
# Import necessary libraries
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report


In [223]:
X4=wine2
y4 = wine1['NumericalLabels']  # --> what you're trying to predict
  # --> the features we will keep to build our model

In [224]:

# Standardize the features (important for SVM)
scaler = StandardScaler()
X4_train = scaler.fit_transform(X_train_resampled)
X4_test = scaler.transform(X_test)

In [225]:
# Initialize the SVM classifier
svm_classifier = SVC(kernel='linear', C=1.0, random_state=42)

In [226]:


# Train the SVM classifier
svm_classifier.fit(X4_train, y_train_resampled)

# Make predictions on the test set
prediction4 = svm_classifier.predict(X4_test)

## RESULT

In [227]:
# Evaluate the performance of the classifier
accuracy4 = accuracy_score(y_test, prediction4)
report4 = classification_report(y_test, prediction4)

# Print the results
print(f"Accuracy: {accuracy4:.2f}")
print("\nClassification Report:\n", report4)
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, prediction4))

Accuracy: 0.51

Classification Report:
               precision    recall  f1-score   support

         0.0       0.56      0.74      0.64       274
         1.0       0.73      0.37      0.49       490
         2.0       0.11      0.69      0.18        29

    accuracy                           0.51       793
   macro avg       0.47      0.60      0.44       793
weighted avg       0.65      0.51      0.53       793


Confusion Matrix:
[[202  56  16]
 [158 179 153]
 [  0   9  20]]


1st try:
Accuracy: 0.70

Classification Report:
               precision    recall  f1-score   support

         0.0       0.64      0.53      0.58       274
         1.0       0.72      0.83      0.77       490
         2.0       0.00      0.00      0.00        29

    accuracy                           0.70       793
   macro avg       0.45      0.45      0.45       793
weighted avg       0.67      0.70      0.68       793


Confusion Matrix:
[[145 129   0]
 [ 82 408   0]
 [  0  29   0]]

2nd try:
Accuracy: 0.52

Classification Report:
               precision    recall  f1-score   support

         0.0       0.56      0.73      0.64       274
         1.0       0.76      0.38      0.51       490
         2.0       0.11      0.72      0.19        29

    accuracy                           0.52       793
   macro avg       0.48      0.61      0.45       793
weighted avg       0.67      0.52      0.54       793


Confusion Matrix:
[[201  52  21]
 [156 188 146]
 [  0   8  21]]

3rd try:
Accuracy: 0.51

Classification Report:
               precision    recall  f1-score   support

         0.0       0.56      0.74      0.64       274
         1.0       0.73      0.37      0.49       490
         2.0       0.11      0.69      0.18        29

    accuracy                           0.51       793
   macro avg       0.47      0.60      0.44       793
weighted avg       0.65      0.51      0.53       793


Confusion Matrix:
[[202  56  16]
 [158 179 153]
 [  0   9  20]]